### PR TITLE
[FEATURE] Task to calculate,store correlation coefficients

### PR DESCRIPTION
**Pull request by stevenzeiler**
DEC-454

### CODE REVIEW

**Comment by stevenzeiler** [comment]
Still need to add a `cron` job to ansible that runs `npm run correlate` every day.

File: tasks/correlate.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    var date = moment().subtract(1, 'day').format('YYYY-MM-DD')
+
+    const score = await CorrelationScore.find({ where: { date: date }})
+
+    if (score) {
+      console.error('Correlation Coefficients already computed for', date)
+    } else {
+      try {
+        const record = await CorrelationCoefficientService.create()
+        console.log('Computd Correlation Coefficients', record.toJSON())
```
Comment (wilsonianb) [inline tasks/correlate.js]: `Computd` -> `Computed`

File: test/services/CorrelationCoefficientService.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+        let validation = await database.Validations.create({
+          ledger_hash: ledgerHash,
+          validation_public_key: validator,
+          reporter_public_key: validator
+        })
+      })
+
+      otherCluster.forEach(async validator => {
+        if (ledgers.indexOf(ledgerHash) % 2 === 0) {
```
Comment (wilsonianb) [inline test/services/CorrelationCoefficientService.js]: What's the purpose of this?

File: test/services/CorrelationCoefficientService.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    CorrelationCoefficientService.compute().then(results => {
+
+      results.forEach(result => {
+
+        assert(result.validation_public_key)
+        assert(result.date_validated)
+        assert(result.denom_validated_ledger, 10)
+
+        if (_.contains(alphaCluster, result.validation_public_key)) {
+          assert.strictEqual(result.num_validated_ledger, 10)
```
Comment (wilsonianb) [inline test/services/CorrelationCoefficientService.js]: This is failing for me periodically with `result.num_validated_ledger` as 9. Is it possible `before` is returning before all the validations are written to the table?

Comment (stevenzeiler) [inline test/services/CorrelationCoefficientService.js]: maybe, I'll give the `before` a callback function and see if that works.

**Comment by wilsonianb** [comment]
It looks like the cron job can be setup in the DockerFile:
http://www.ekito.fr/people/run-a-cron-job-with-docker/

**Comment by wilsonianb** [comment]
Maybe the cron jobs should be run individual containers to follow the "one process per container" principle.
https://docs.docker.com/articles/dockerfile_best-practices/

File: package.json [hunk 4]
```diff
@@ -38,7 +38,8 @@
   "scripts": {
     "debug": "node debug app.js",
     "start": "node app.js",
-    "test": "BASIC_AUTH_USER=user BASIC_AUTH_PASS=pass mocha test/**/* --compilers=js:babel/register --timeout 15000"
+    "test": "BASIC_AUTH_USER=user BASIC_AUTH_PASS=pass mocha test/**/* --compilers=js:babel/register --timeout 15000",
+    "correlate": "grunt correlate"
```
Comment (wilsonianb) [inline package.json]: I think grunt needs to be installed in the Dockerfile with:
`RUN npm install -g grunt-cli`

**Comment by stevenzeiler** [comment]
:+1: The cron job belongs in a separate pull request.

File: test/services/CorrelationCoefficientService.js (hunk 3 above)
Comment (wilsonianb) [inline test/services/CorrelationCoefficientService.js]: Can you push the fix to the `before`?

Comment (wilsonianb) [inline test/services/CorrelationCoefficientService.js]: Maybe replace the `forEach`s with a `for` loop?

Comment (wilsonianb) [inline test/services/CorrelationCoefficientService.js]: Maybe replace the `forEach`s with `for` loops?
http://jakearchibald.com/2014/es7-async-functions/

Comment (stevenzeiler) [inline test/services/CorrelationCoefficientService.js]: That works!

**Comment by wilsonianb** [comment]
LGTM :+1:  Let's wait to merge until deploys are working again

**Comment by stevenzeiler** [comment]
okay

**Pull request merged**